BaseLib/ConfigTreeNew-impl.h

```diff
@@ -133,15 +116,70 @@ checkConfParam(std::string const& param, Ch const* value) const
     }
 }
 
+template<typename T>
+T
+ConfigTreeNew::
+getValue() const
+{
+    if (_have_read_data) {
+        error("The data of this subtree has already been read.");
+    }
+
+    _have_read_data = true;
+
+    if (auto v = _tree->get_value_optional<T>()) {
+        return *v;
+    } else {
+        error("Value `" + shortString(_tree->data())
+              + "' is not convertible to the desired type.");
+    }
+}
+
+template<typename T>
+T
+ConfigTreeNew::
+getConfAttribute(std::string const& attr) const
+{
+    if (auto a = getConfAttributeOptional<T>(attr))
+        return *a;
+
+    error("Did not find XML attribute with name \"" + attr + "\".");
+}
+
+template<typename T>
+boost::optional<T>
+ConfigTreeNew::
+getConfAttributeOptional(std::string const& attr) const
+{
+    checkUniqueAttr(attr);
+    auto& ct = markVisited<T>(attr, true, true);
+
+    if (auto attrs = _tree->get_child_optional("<xmlattr>")) {
+        if (auto a = attrs->get_child_optional(attr)) {
+            ++ct.count; // count only if attribute has been found
+            if (auto v = a->get_value_optional<T>()) {
+                return v;
+            } else {
+                error("Value for XML attribute \"" + attr + "\" `"
+                      + shortString(a->data())
+                      + "' not convertible to the desired type.");
+            }
+        }
+    }
+
+    return boost::none;
+}
 
 template<typename T>
 ConfigTreeNew::CountType&
 ConfigTreeNew::
-markVisited(std::string const& key, bool peek_only) const
+markVisited(std::string const& key, bool const is_attr,
+            bool const peek_only) const
```

Comment:
Maybe it is better to introduce an enum for `is_attr` parameter with the values like `ATTR` and `PARAM`. This would result in better code where `markVisited` is used, _e.g._:
```c++
auto& ct = markVisited<T>(attr, ATTR, true); //line 155 above
markVisited<T>(param, PARAM, true); // line 72 above
```
The implementation would not change when choosing ATTR=1 and PARAM=0. And it should result in identical assembler code.